Support compressed files in data_manager_fetch_genome_all_fasta_dbkeys

mvdbeek opened the pull request:
It's hard these days to find uncompressed fasta downloads anywhere ... so this PR
will check if input files (download or history/directory, except for symlinks for obvious reasons) look like they are compressed and uncompress them before writing them out.

Also minor tweaks to the help text, use of CDATA tag, avoid `<command interpreter="python">`
and py2/py3 compatibility.

I have uploaded the manager under my name to the TTS if someone would like to test https://testtoolshed.g2.bx.psu.edu/view/mvdbeek/data_manager_fetch_genome_dbkeys_all_fasta_compressed/0738889ff4d2 

I couldn't quite figure out if/how I can use planemo to test a data manager, but I have uploaded a workflow [here](https://gist.github.com/mvdbeek/9cee730483a68b36d9bf23c3eb3897af) that tests most of the functionality.

nsoranzo commented on data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py:
```diff
@@ -175,155 +200,163 @@ def _sort_fasta_custom( fasta_filename, params ):
-    ext = None
-    ucsc_chrom_fa_filename = None
+    path_contents = _get_files_in_ftp_path(ftp, ucsc_path)
+    ftp.close()
```
It seems from Python doc and Google search, that `ftp.quit()` should be the proper way to close the connection, but not really sure.

mvdbeek replied:
```
In [2]: ?FTP.close
Signature: FTP.close(self)
Docstring: Close the connection without assuming anything about it.
File:      /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ftplib.py
Type:      instancemethod

In [3]: ?FTP.quit
Signature: FTP.quit(self)
Docstring: Quit, and close the connection.
File:      /usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ftplib.py
Type:      instancemethod
```
Seems quit is more gentle? Will do that.

mvdbeek commented:
Anyone willing to 👍 ? @bgruening @blankenberg ?

bgruening commented on data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py:
```diff
@@ -405,11 +439,16 @@ def main():
+        cleanup_before_exit(tmp_dir)
     #save info to json file
-    open( filename, 'wb' ).write( dumps( data_manager_dict ) )
+    open( filename, 'w' ).write( dumps( data_manager_dict ) )
```
Why do you change it here and use `wb` above. `wb` does nothing on Unix like operating systems afaik. Just wondering.

mvdbeek replied:
I'll check that. IIRC python3 didn't want to dump the json file when the file was opened in 'wb' mode.

bgruening commented on data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.xml:
```diff
@@ -1,86 +1,85 @@
+                </repeat>
+                <conditional name="handle_not_listed">
+                    <param name="handle_not_listed_selector" type="select" label="How to handle non-specified Identifiers">
+                        <option value="discard" selected="True">Discard</option>
```
Why is this a conditional? I know it not your fault but maybe we can improve things here a little bit.

mvdbeek replied:
You're right, I think we can simply remove the conditional.

bgruening commented on data_managers/data_manager_fetch_genome_dbkeys_all_fasta/data_manager/data_manager_fetch_genome_all_fasta_dbkeys.py:
```diff
@@ -175,155 +200,163 @@ def _sort_fasta_custom( fasta_filename, params ):
     UCSC_DOWNLOAD_PATH = '/goldenPath/%s/bigZips/'
-    COMPRESSED_EXTENSIONS = [ ( '.tar.gz', _get_stream_readers_for_tar ), ( '.tar.bz2', _get_stream_readers_for_tar ), ( '.zip', _get_stream_readers_for_zip ), ( '.fa.gz', _get_stream_readers_for_gzip ), ( '.fa.bz2', _get_stream_readers_for_bz2 ) ]
-    
+    COMPRESSED_EXTENSIONS = ['.tar.gz', '.tar.bz2', '.zip', '.fa.gz', '.fa.bz2']
```
tgz?

mvdbeek replied:
Safe addition, not sure they really have that extension, but I'll add it.

mvdbeek commented:
@bgruening I removed the conditional and added .tgz to the list of files downloadable from UCSC.
About the binary mode: I played around with this some more, python3 will refuse to write bytes (such as received by downloading files or reading in intermediary fasta files) to files open only in `'w'` mode. So I also tried decoding the incoming bytestreams and encoding before feeding this back to the compressed stream handler, but there was always something wrong in these modules in conjunction with StringIO or BytesIO.
I think this is currently the easiest solution to support both python3 and python2 (maybe not the best, but I don't think this would be a regression from the previous code).

bgruening commented:
Sounds good to me! Thanks!

mvdbeek commented:
Thanks @bgruening ! @martenson can you push it pleaseee?

martenson commented:
@mvdbeek done https://toolshed.g2.bx.psu.edu/view/devteam/data_manager_fetch_genome_dbkeys_all_fasta/86fa71e9b427

mvdbeek commented:
Thanks @martenson
